File: obs_replaykit/audio.py
"""enumerate windows audio devices. obs stores endpoint ids in mmdevice format -- {0.0.1.00000000}.<endpoint-guid> for capture, {0.0.0.00000000}.<endpoint-guid> for render. friendly names live under each endpoints Properties subkey."""

import winreg
from dataclasses import dataclass
from typing import List, Optional


# pkey_device_friendlyname in the mmdevices property store.
_PKEY_FRIENDLY_NAME = "{a45c254e-df1c-4efd-8020-67d146a850e0},2"

_CAPTURE_KEY = r"SOFTWARE\Microsoft\Windows\CurrentVersion\MMDevices\Audio\Capture"
_RENDER_KEY  = r"SOFTWARE\Microsoft\Windows\CurrentVersion\MMDevices\Audio\Render"

# devicestate bits. only the low nibble carries the documented mmdevice states; some virtual-audio drivers (voicemeeter etc.) set extra high bits.
_STATE_ACTIVE = 0x1

# literal string obs uses for 'follow the system defualt device'.
DEFAULT_DEVICE_ID = "default"

DEFAULT_DEVICE_NAME = "Default (system)"


@dataclass(frozen=True)
class AudioDevice:
    """An OBS-format audio endpoint."""
    device_id: str  # obs-format device id ({0.0.1.00000000}.<guid>)
    name: str  # friendly name (e.g. 'microphone (realtek)')


def _to_obs_id(endpoint_guid: str) -> str:
    return f"{{0.0.1.00000000}}.{endpoint_guid}"


def _to_render_obs_id(endpoint_guid: str) -> str:
    return f"{{0.0.0.00000000}}.{endpoint_guid}"


def find_render_endpoint(friendly_name: str) -> Optional[str]:
    """OBS-format render device id matching friendly_name (case-insensitive exact). Used to write the fresh OBS Stream Audio id straight into basic.ini so OBS picks the right monitoring sink on first launch."""
    needle = friendly_name.lower()
    try:
        render = winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, _RENDER_KEY)
    except OSError:
        return None

    with render:
        i = 0
        while True:
            try:
                endpoint = winreg.EnumKey(render, i)
            except OSError:
                break
            i += 1

            try:
                with winreg.OpenKey(render, endpoint) as ep:
                    state, _ = winreg.QueryValueEx(ep, "DeviceState")
            except OSError:
                continue
            if not (state & _STATE_ACTIVE):
                continue

            try:
                with winreg.OpenKey(render, fr"{endpoint}\Properties") as props:
                    name, _ = winreg.QueryValueEx(props, _PKEY_FRIENDLY_NAME)
            except OSError:
                continue

            if name.lower() == needle:
                return _to_render_obs_id(endpoint)

    return None


def _list_render_endpoints() -> List[AudioDevice]:
    """Active render endpoints known to Windows, in OBS id format."""
    devices: List[AudioDevice] = []
    try:
        render = winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, _RENDER_KEY)
    except OSError:
        return devices

    with render:
        i = 0
        while True:
            try:
                endpoint = winreg.EnumKey(render, i)
            except OSError:
                break
            i += 1

            try:
                with winreg.OpenKey(render, endpoint) as ep:
                    state, _ = winreg.QueryValueEx(ep, "DeviceState")
            except OSError:
                continue
            if not (state & _STATE_ACTIVE):
                continue

            try:
                with winreg.OpenKey(render, fr"{endpoint}\Properties") as props:
                    name, _ = winreg.QueryValueEx(props, _PKEY_FRIENDLY_NAME)
            except OSError:
                name = endpoint

            devices.append(AudioDevice(_to_render_obs_id(endpoint), name))

    return devices


def _replaykit_monitor_rank(name: str) -> Optional[int]:
    """Rank only the ReplayKit/VB-CABLE render sink; never match normal speakers."""
    lower = name.strip().lower()
    if "surround" in lower or "16ch" in lower:
        return None
    if lower == "obs stream audio":
        return 0
    if lower.startswith("obs stream audio") and "loopback" not in lower:
        return 10
    if lower.startswith("cable input"):
        return 20
    return None


def find_replaykit_monitoring_endpoint() -> Optional[AudioDevice]:
    """Best active render endpoint for OBS monitoring, before or after ReplayKit's VB-CABLE rename."""
    candidates: List[tuple[int, str, AudioDevice]] = []
    for device in _list_render_endpoints():
        rank = _replaykit_monitor_rank(device.name)
        if rank is not None:
            candidates.append((rank, device.name.lower(), device))
    if not candidates:
        return None
    candidates.sort(key=lambda item: (item[0], item[1]))
    return candidates[0][2]


def list_microphones() -> List[AudioDevice]:
    """active capture devices known to windows, alphabetized."""
    devices: List[AudioDevice] = []
    try:
        cap = winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, _CAPTURE_KEY)
    except OSError:
        return devices

    with cap:
        i = 0
        while True:
            try:
                endpoint = winreg.EnumKey(cap, i)
            except OSError:
                break
            i += 1

            # active devices only (low bit set).
            try:
                with winreg.OpenKey(cap, endpoint) as ep:
                    state, _ = winreg.QueryValueEx(ep, "DeviceState")
            except OSError:
                continue
            if not (state & _STATE_ACTIVE):
                continue

            try:
                with winreg.OpenKey(cap, fr"{endpoint}\Properties") as props:
                    name, _ = winreg.QueryValueEx(props, _PKEY_FRIENDLY_NAME)
            except OSError:
                name = endpoint

            devices.append(AudioDevice(_to_obs_id(endpoint), name))

    devices.sort(key=lambda d: d.name.lower())
    return devices

File: ReplayKit/ReplayKitSetup/Audio.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.Win32;

namespace ReplayKitSetup
{
    // enumerate windows audio devices. obs stores endpoint ids in mmdevice format -- {0.0.1.00000000}.<endpoint-guid> for capture, {0.0.0.00000000}.<endpoint-guid> for render. friendly names live under each endpoints properties subkey. ported from obs_replaykit/audio.py.
    public static class Audio
    {
        // pkey_device_friendlyname in the mmdevices property store.
        private const string PkeyFriendlyName = "{a45c254e-df1c-4efd-8020-67d146a850e0},2";

        private const string CaptureKey = @"SOFTWARE\Microsoft\Windows\CurrentVersion\MMDevices\Audio\Capture";
        private const string RenderKey = @"SOFTWARE\Microsoft\Windows\CurrentVersion\MMDevices\Audio\Render";

        // devicestate bits. only the low nibble carries the documented mmdevice states; some virtual-audio drivers (voicemeeter etc.) set extra high bits.
        private const int StateActive = 0x1;

        // literal string obs uses for "follow the system defualt device".
        public const string DEFAULT_DEVICE_ID = "default";
        public const string DEFAULT_DEVICE_NAME = "Default (system)";

        public sealed class AudioDevice
        {
            public string DeviceId { get; }
            public string Name { get; }

            public AudioDevice(string deviceId, string name)
            {
                DeviceId = deviceId;
                Name = name;
            }
        }

        private static string ToObsId(string endpointGuid) => "{0.0.1.00000000}." + endpointGuid;
        private static string ToRenderObsId(string endpointGuid) => "{0.0.0.00000000}." + endpointGuid;

        private static IEnumerable<(string Endpoint, string Name)> ActiveEndpoints(string rootKeyPath)
        {
            using (var root = Registry.LocalMachine.OpenSubKey(rootKeyPath))
            {
                if (root == null) yield break;

                foreach (var endpoint in root.GetSubKeyNames())
                {
                    int state;
                    using (var ep = root.OpenSubKey(endpoint))
                    {
                        if (ep == null) continue;
                        var stateValue = ep.GetValue("DeviceState");
                        if (stateValue == null) continue;
                        state = Convert.ToInt32(stateValue);
                    }
                    if ((state & StateActive) == 0) continue;

                    string name = endpoint;
                    using (var props = root.OpenSubKey(endpoint + @"\Properties"))
                    {
                        var nameValue = props?.GetValue(PkeyFriendlyName) as string;
                        if (!string.IsNullOrEmpty(nameValue)) name = nameValue;
                    }

                    yield return (endpoint, name);
                }
            }
        }

        // obs-format render device id matching friendlyName (case-insensitive exact). used to write the fresh obs stream audio id straight into basic.ini so obs picks the right monitoring sink on first launch. unlike ActiveEndpoints, an endpoint whose friendly name cant be read is skipped rather than matched by its raw guid -- matches the python originals continue-on-failure behavior exactly.
        public static string FindRenderEndpoint(string friendlyName)
        {
            string needle = friendlyName.ToLowerInvariant();
            using (var render = Registry.LocalMachine.OpenSubKey(RenderKey))
            {
                if (render == null) return null;

                foreach (var endpoint in render.GetSubKeyNames())
                {
                    using (var ep = render.OpenSubKey(endpoint))
                    {
                        var stateValue = ep?.GetValue("DeviceState");
                        if (stateValue == null) continue;
                        if ((Convert.ToInt32(stateValue) & StateActive) == 0) continue;
                    }

                    using (var props = render.OpenSubKey(endpoint + @"\Properties"))
                    {
                        var name = props?.GetValue(PkeyFriendlyName) as string;
                        if (name == null) continue;
                        if (string.Equals(name, needle, StringComparison.OrdinalIgnoreCase)) return ToRenderObsId(endpoint);
                    }
                }
            }
            return null;
        }

        private static List<AudioDevice> ListRenderEndpoints()
        {
            return ActiveEndpoints(RenderKey).Select(e => new AudioDevice(ToRenderObsId(e.Endpoint), e.Name)).ToList();
        }

        // rank only the replaykit/vb-cable render sink; never match normal speakers.
        private static int? ReplaykitMonitorRank(string name)
        {
            string lower = name.Trim().ToLowerInvariant();
            if (lower.Contains("surround") || lower.Contains("16ch")) return null;
            if (lower == "obs stream audio") return 0;
            if (lower.StartsWith("obs stream audio") && !lower.Contains("loopback")) return 10;
            if (lower.StartsWith("cable input")) return 20;
            return null;
        }

        // best active render endpoint for obs monitoring, before or after replaykits vb-cable rename.
        public static AudioDevice FindReplaykitMonitoringEndpoint()
        {
            var candidates = new List<(int Rank, string LowerName, AudioDevice Device)>();
            foreach (var device in ListRenderEndpoints())
            {
                var rank = ReplaykitMonitorRank(device.Name);
                if (rank.HasValue) candidates.Add((rank.Value, device.Name.ToLowerInvariant(), device));
            }
            if (candidates.Count == 0) return null;
            candidates.Sort((a, b) =>
            {
                int cmp = a.Rank.CompareTo(b.Rank);
                return cmp != 0 ? cmp : string.CompareOrdinal(a.LowerName, b.LowerName);
            });
            return candidates[0].Device;
        }

        // active capture devices known to windows, alphabetized.
        public static List<AudioDevice> ListMicrophones()
        {
            var devices = ActiveEndpoints(CaptureKey).Select(e => new AudioDevice(ToObsId(e.Endpoint), e.Name)).ToList();
            devices.Sort((a, b) => string.Compare(a.Name, b.Name, StringComparison.OrdinalIgnoreCase));
            return devices;
        }
    }
}
